### ISSUE TITLE
pmap strips `Date`

### ISSUE THREAD

**Issue by AkhilNairAmey** [tidyverse/purrr#358]
As title - `purrr::pmap` seems to strip date class.

``` r
print_date = function(b) b

b = as.Date("2012-03-22")

purrr::pmap(list(b), print_date)
#> [[1]]
#> [1] 15421
```

Not sure if this is intended? Caught me out though :)

**Comment by david-jankoski** [comment]
there were a couple of similar discussions around `map` and dates around here. this workaround does the job for me. 
```
print_date = function(b) as.Date(b, origin = "1970-01-01")
b = as.Date("2012-03-22")
purrr::pmap(list(b), print_date)

[[1]]
[1] "2012-03-22"
``` 
hth.

**Comment by AkhilNairAmey** [comment]
Thanks @david-jankoski - That's what I'm currently doing. Just wanted to make sure it was known, I admitted didn't look around to see if anything similar has been asked.

**Comment by mungojam** [comment]
I hit the same problem with pwalk

**Comment by huftis** [comment]
The bug is not specific to dates; `pmap()` removes all classes. Note that `map()` doesn’t have this bug. Simple reprex:

``` r
library(purrr)
x = "test"
class(x) = "foo"
pmap(list(x), identity)
#> [[1]]
#> [1] "test"
map(list(x), identity)
#> [[1]]
#> [1] "test"
#> attr(,"class")
#> [1] "foo"
```

And as @mungojam mentioned, `pwalk()` also has this bug:

```r
> pwalk(list(x), ~ print(class(.)))
# [1] "character" # should have been "foo" ...
```

**Comment by lionel-** [comment]
This is more of a known limitation rather than a bug, but we have plans to fix this.

**Comment by mungojam** [comment]
I found today that flatten does this too and removed my Date class as well as doing the intended un-listing. Perhaps that is intentional?

**Comment by mungojam** [comment]
I found today that `flatten` does this too and removed my Date class as well as doing the intended un-listing. Perhaps that is intentional?

**Comment by lionel-** [comment]
It is not intentional we've just been too busy with tidy eval and other projects :(

**Comment by mungojam** [comment]
No worries, I'm enjoying all your other work :)

**Comment by hadley** [comment]
The root cause is this behaviour:

``` r
x <- as.Date("2012-03-22")
x[[1]]
#> [1] "2012-03-22"
list(x)[[c(1,1)]]
#> [1] 15421
```

This seems like a base R problem?

**Comment by yutannihilation** [comment]
> This seems like a base R problem?

It seems so to me; `[[`, or `do_subset2()`, tries to dispatch the method based only on the first argument: https://github.com/wch/r-source/blob/58aec3ee94f2a0769974520f6cb2d99c49fb829f/src/main/subset.c#L894-L914

If it fails, the default implementation of subset2 will be used for all subsets, but, IIUC, it doesn't care about the classes or attributes: https://github.com/wch/r-source/blob/58aec3ee94f2a0769974520f6cb2d99c49fb829f/src/main/subset.c#L1029-L1056

**Comment by czeildi** [comment]
Related: I'd love to see a `map_date` and `map_datetime` functions. My use case: there are some non-vectorized datetime related functions, e.g. `lubridate::with_tz` and I need map to create a new column with them in a data frame. I only found a workaround to arrive a plain datetime column.

``` r
library(purrr)
library(lubridate)
library(tibble)
library(dplyr)

df <- tibble(
  dttm_orig = ymd_hms(c("2018-04-01 08:00:00", "2017-03-02 03:00:00")),
  tz = c("UTC", "CET"),
  sales_amount = c(34, 56)
)

mutate(df, purchase_time = map2(dttm_orig, tz, ~with_tz(.x, .y)))
#> # A tibble: 2 x 4
#>   dttm_orig           tz    sales_amount purchase_time
#>   <dttm>              <chr>        <dbl> <list>       
#> 1 2018-04-01 08:00:00 UTC           34.0 <dttm [1]>   
#> 2 2017-03-02 03:00:00 CET           56.0 <dttm [1]>

mutate(df, purchase_time = map2_chr(dttm_orig, tz, ~as.character(with_tz(.x, .y))) %>% ymd_hms())
#> # A tibble: 2 x 4
#>   dttm_orig           tz    sales_amount purchase_time      
#>   <dttm>              <chr>        <dbl> <dttm>             
#> 1 2018-04-01 08:00:00 UTC           34.0 2018-04-01 08:00:00
#> 2 2017-03-02 03:00:00 CET           56.0 2017-03-02 04:00:00
```

Created on 2018-05-01 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

I realize this might not be top priority however I would love to vote up an issue dedicated to this specifically.

**Comment by hadley** [comment]
To resolve this issue, we need to make sure to file a bug in R's bugzilla.

**Comment by yutannihilation** [comment]
Just out of curiosity, why is this not just about replacing `.l[[c(j, i)]]` with `.l[[j]][[i]]` here?

https://github.com/tidyverse/purrr/blob/d0a808186e820fb637affe4d92cef2c7bf3cf6bd/src/map.c#L189-L192

(I'm afraid this is not a bug, since it's reasonable that the method dispatch happens only once for one method.)

**Comment by burchill** [comment]
It might already be known, but calling `as_vector()` on a list of dates _also_ strips `Date`.

**Comment by yutannihilation** [comment]
@burchill Though it seems similar, I feel it's a separate issue. Maybe you should file a new issue. (BTW, I guess what you need is not `as_vector()`, but something like `flatten_vec()`. I expect `_vec`-family will be implemented in future. e.g. `map_vec()` on #499)

**Comment by hadley** [comment]
@mikmart oh good idea. Want to do a PR?

**Comment by yutannihilation** [comment]
Thanks for the PR, @mikmart!
